### ISSUE TITLE
Memory leak when hitting a site over https with invalid certificates.

### ISSUE THREAD

**Issue by gcharnock** [nodejs/node-v0.x-archive#25616]
We've been tracking down a memory leak that's been occurring in some of our node apps for a few days, and we've been able to produce a test case that rapidly leaks memory whenever there is a certificate error. This eventually leads to the node process crashing when deployed. The test case presented here uses a server with a self signed certificate.

Client that memory leaks:

```
var https = require("https");

function tick() {
    var req = https.request({
        hostname: "localhost",
        port: 8000,
        path: "/",
        method: "GET"
    }, function(response){
        throw new Error("This test case should hit something with an invalid certificate, execution should not reach here.");
    });
    req.end();
    //make sure something is pulling data out, just to be use.
    req.on("data", function(){})
    req.on('error', function(err) {
        console.log("request made, err was " + err);
    });
    global.gc(); //force the gc to run to increase signal to noise ratio.
}

setInterval(tick, 10);
```

Server:

```

var https = require('https');
var fs = require('fs');

var options = {
  key: fs.readFileSync('./key.pem'),
  cert: fs.readFileSync('./cert.pem')
};

https.createServer(options, function (req, res) {
  res.writeHead(200);
  res.end("hello world\n");
}).listen(8000);

```

Script to run both:

```
openssl req -x509 -nodes -newkey rsa:2048 -keyout key.pem -out cert.pem -days 2000
node server.js & node --expose-gc main.js
```

The node version is v0.12.5

**Comment by PaulBGD** [comment]
+1, thank you for tracking this down

**Comment by ehacke** [comment]
+1

**Comment by brendanashworth** [comment]
I'm sorry about the lack of a response on this. Thanks for the easily reproduceable test case - I can confirm this occurs on 0.12.7, where the heap grows uncontrollably, but it does _not_ occur on master (v4). I'll be looking into this but if this is an issue for any of you, you can probably upgrade to v4 and not have this issue occur.

**Comment by brendanashworth** [comment]
Got it: this particular leak was fixed in 2ccc8f3970cd1ed78585be37c8d912f3ccf56cd0 (also, e74b5d278ca94bf27d16b3fb94cedf4c6fcac3ac was with it, that might be backported to.) Ref to https://github.com/nodejs/node/pull/1244 and https://github.com/nodejs/node/issues/1075.

cc @indutny @nodejs/lts I'd consider these two commits for a backport to v0.12. This memory leak does not occur on v0.10 or master.
